File: algorithms/lc/1047.Remove-All-Adjacent-Duplicates-In-String.py
# this is like a version of the valid parens

def removeDuplicates(S):
    stack = []
    for char in S:
        # if there is a stack to pop check if the top is the char
        if stack and stack[-1] == char:
            stack.pop()
        #otherwise add to the top of the stack
        else:
            stack.append(char) # [a]
            print(stack)
        # print(stack)
    return ''.join(stack)
print(removeDuplicates('abbaca'))

"""
['a']
['a', 'b']
['c']
['c', 'a']
ca

"""

File: algorithms/lc/1047.Remove-All-Adjacent-Duplicates-In-String.js
function removeDuplicates(S) {
    const stack = [];
    for (char of S) {
        console.log(stack);
        if (stack.length > 0 && stack[stack.length - 1] === char) {
            stack.pop();
        } else {
            stack.push(char);
        }
    }
    return stack.join('');
}

console.log(removeDuplicates('abbaca'))
/* 
[]
[ 'a' ]
[ 'a', 'b' ]
[ 'a' ]
[]
[ 'c' ]
ca
*/
